src/Microsoft.AspNetCore.OData/Routing/Conventions/OperationRoutingConvention.cs

```diff
@@ -65,31 +65,22 @@ protected void ProcessOperations(ODataControllerActionContext context, IEdmEntit
                 return;
             }
 
-            // OperationNameOnCollectionOfEntityType
-            string operationName = SplitActionName(actionName, out string cast, out bool isOnCollection);
-
+            bool isOnCollection = false;
             IEdmEntityType castTypeFromActionName = null;
-            if (cast != null)
+
+            IEdmOperation[] candidates = FindCandidates(context, actionName);
+            if (candidates.Length == 0)
             {
-                if (cast.Length == 0)
-                {
-                    // Early return for the following cases:
-                    // - {OperationName}On
-                    // - {OperationName}OnCollectionOf
-                    return;
-                }
+                // If we can't find any Edm operation using the action name directly,
+                // Let's split the action name and use part of it to search again.
+                candidates = FindCandidates(context, entityType, actionName, out castTypeFromActionName, out isOnCollection);
+            }
 
-                castTypeFromActionName = entityType.FindTypeInInheritance(context.Model, cast, context.Options?.RouteOptions?.EnableActionNameCaseInsensitive == true) as IEdmEntityType;
-                if (castTypeFromActionName == null)
-                {
-                    return;
-                }
+            if (candidates == null || candidates.Length == 0)
```

Comment:
Can we make it so that none of the `FindCandidates` overloads ever return `null`, and instead make it return an empty array (for instance, using `Array.Empty<T>`)?

Handling null collections like this is usually ill-advised unless there is a very clear semantic value difference between null and empty (which I don't think is the case here).

We could then simplify this back to:
```suggestion
            if (candidates.Length == 0)
```

Reply:
#Resolved.